### PR TITLE
enforce database env prefix

### PR DESCRIPTION
**Pull request by lunika**
## Purpose

To prevent having the same prefix between different environments, we have created a mapping where all environment are defined with a corresponding prefix.

## Proposal

Create a mapping between and their prefixes and use them to generate database credentials.
Also added in this PR mongodb admin password generation.

### CODE REVIEW

File: plugins/filter/merge.py [hunk 1]
```diff
@@ -88,7 +88,18 @@ def merge_with_database(base, database, app_name, customer, environment):
...
+    mapping_env = {
+        "ci": "c",
+        "development": "d",
+        "feature": "f",
+        "production": "p",
+        "staging": "s",
+        "preprod": "t",
+    }
+
+    env_type = mapping_env.get(environment) if environment in mapping_env else environment[0]
```
Comment (sampaccoud) [inline plugins/filter/merge.py]: You can write it like this
```python
 env_type = mapping_env.get(environment, environment[0])
```

This being said, I think this could create unexpected issues if the environment is not in your mapping but starts with the same letter as one of your environments? 

Should we make it configurable and raise an issue if the environment is not known?

Comment (lunika) [reply plugins/filter/merge.py]: Yes this can lead to an environment having the same letter. I don't know how ta make it configurable. If we want to make it configurable we have to inject in the filter I think (I will check the documentation) and I'm not really comfortable with this.

Comment (jmaupetit) [reply plugins/filter/merge.py]: Yes: it should be configurable at the Ansible level and injected while calling the filter.

```yaml
# groups_vars/all/main.yml
env_types:
  - name: "development"
    code: "d"
  - name: "staging"
    code: "s"
# [...]
```

Usage of the merge plugin should be:

```yaml
# tasks/create_app_database_vault.yml
databases: "{{ databases | default({}) | merge_with_database(item, app.name, customer, env_type, env_code) }}"
```

Comment (jmaupetit) [reply plugins/filter/merge.py]: or (better):

```yaml
# tasks/create_app_database_vault.yml
databases: "{{ databases | default({}) | merge_with_database(item, app.name, customer, env_types | json_query(name=env_type) }}"
```

The latest parameter will be a `dict` with the environment name and code.

Comment (lunika) [reply plugins/filter/merge.py]: :+1:

File: group_vars/all/main.yml [hunk 2]
```diff
@@ -3,6 +3,25 @@ env_type: development  # default to be overriden on command line
 customer: eugene  # default to be overriden on command line
 project_name: "{{ env_type }}-{{ customer }}"
 
+# env_types
+# for each environment, a code is associated. This code will be used
+# to generate databases name and user with this pattern: {code}_{customer}_{app}
+# for example for the application marsha in development and the user eugene a database name will be d_eugene_marsha
+# If you need more environment you can extends this list respecting name and code keys and a unique code value
```
Comment (sampaccoud) [inline group_vars/all/main.yml]: s/extends/extend

File: tasks/create_app_database_vault.yml [hunk 3]
```diff
@@ -14,6 +14,6 @@
 
 - name: Append database to databases vault
   set_fact:
-    databases: "{{ databases | default({}) | merge_with_database(item, app.name, customer, env_type) }}"
+    databases: "{{ databases | default({}) | merge_with_database(item, app.name, customer, env_types | json_query(\"[?name=='\" ~ env_type ~ \"'] | [0]\")) }}"
```
Comment (sampaccoud) [inline tasks/create_app_database_vault.yml]: I would not mind a comment explaining what the json query is selecting in plain english if its possible. I wanted to propose something but it looked like : selected the env_type from the env_types that matches the env_type :-(
A sign that the name of the `env_types` list is not optimal?

File: apps/edxapp/vars/vault/main.yml.j2 [hunk 4]
```diff
@@ -12,7 +12,7 @@ MYSQL_PASSWORD: {{ mysql_credentials.password }}
 
 # MongoDB
 {% set mongodb_credentials = databases.mongodb | json_query("[?release=='" ~ edxapp_mongodb_version ~ "'].databases | [0][?application=='edxapp'].{user: user, password: password, name: name} | [0]") %}
-MONGODB_ADMIN_PASSWORD: password
+MONGODB_ADMIN_PASSWORD: "{{ lookup('password','/dev/null length=24') }}"
```
Comment (jmaupetit) [inline apps/edxapp/vars/vault/main.yml.j2]:
```suggestion
MONGODB_ADMIN_PASSWORD: "{{ lookup('password', '/dev/null length=24') }}"
```

File: apps/forum/vars/vault/main.yml.j2 [hunk 5]
```diff
@@ -4,7 +4,7 @@
 
 # MongoDB
 {% set mongodb_credentials = databases.mongodb | json_query("[?release=='" ~ forum_mongodb_version ~ "'].databases | [0][?application=='forum'].{user: user, password: password, name: name} | [0]") %}
-MONGODB_ADMIN_PASSWORD: password
+MONGODB_ADMIN_PASSWORD: "{{ lookup('password','/dev/null length=24') }}"
```
Comment (jmaupetit) [inline apps/forum/vars/vault/main.yml.j2]:
```suggestion
MONGODB_ADMIN_PASSWORD: "{{ lookup('password', '/dev/null length=24') }}"
```

File: tasks/create_app_database_vault.yml (hunk 3 above)
Comment (jmaupetit) [reply tasks/create_app_database_vault.yml]: Would you prefer `environments`? And the related comment:

> Select the environment matching the current `env_type`

File: group_vars/all/main.yml (hunk 2 above)
Comment (jmaupetit) [inline group_vars/all/main.yml]: I would propose the following comment:

```yaml
# Environments
#
# Each environment is defined by a name and a code. These environments can be
# extended given your needs, but they should respect the following conventions:
#
# 1. the current env_type should match one of those environments name,
# 2. an environment code should be unique among all environments.
#
# You should note that the environment code will be used to generate database
# credentials with the following pattern: 
#
# {{ environment.code }}_{{ customer }}_{{ app.name }}
#
# For example, the database name for the marsha application deployed in
# development environment for the eugene customer will be "d_eugene_marsha".
```

**Comment by lunika** [comment]
This is out of the scope of this PR but we could maybe rename the `validate_customer_name` task to a more generic one contain all the verification needed to run a playbook. We could add a verification that `environments` contains the current `env_type`. And verifications can be added later. WDYT ?

**Comment by jmaupetit** [comment]
You can postpone this improvement, but don't forget to declare an issue for that.

**Review by jmaupetit — APPROVED** [review]
